docs/repositories.md

```diff
@@ -124,8 +124,7 @@ default = false
 secondary = false
 ```
 
-Any package source not marked as `secondary` will take precedence over [PyPI](https://pypi.org).
-
+The resolution order for searching packages is the following. First, the optional `default` package source is considered. Second, the `primary` (non-default, non-secondary) package sources are considered in order of appearance in your [`pyproject.toml`]({{< relref "pyproject" >}}) file. Third, unless another `default` package source has been configured, the [PyPI](https://pypi.org) repository is considered. Note that all non-secondary packages take precedence over [PyPI](https://pypi.org). Finally, the `secondary` package sources are considered in order of appearance.
```

Comment:
This isn't accurate -- `default` is mentioned twice and the first mention is incorrect. In addition, I would double quote (`"`) "primary" (instead of writing it as `primary`) to make it clear that it's just the internal nickname and not a TOML-level option.

Reply:
It might be inaccurate, but I don't think the first mention is incorrect. IIUC, the second mention is not always correct.

According to https://github.com/python-poetry/poetry/blob/23c4f9b34bf897948b85b44e351d40d064f7efb5/src/poetry/factory.py#L144-L153

- If there is an explicit default, PyPI is not used.
- If there is no explicit default
  - and you don't have any primary repositories, PyPI is the default.
  - and you have primary repositories there is no default. PyPI is **not** the default, but the  secondary with the lowest priority.

The last case is not described properly.

Reply:
I think we're both agreeing but in different ways... Here is how I would describe the resolution logic:

* First, "primary" sources are considered in the order they are defined. These are entries in the pyproject.toml without `default = true` or `secondary = true`. This is equivalent to pip's `--extra-index-url`.
* Second, the `default` source is considered. If the default source is not overidden in pyproject.toml, this is PyPI. This is equivalent to pip's `--index-url`.
* Finally, all `secondary` sources are considered in the order they are defined. This is a Poetry invention.

Reply:
Not entirely correct: If there is a default source (explicit or implicit), the default source is always the first primary source. (The default source is always considered first.)

Reply:
I don't disagree on a source code level, but I am referring to the practical behavior an end user sees:

Given a primary source, `click` and friends will be locked from the primary source:
```toml
[tool.poetry.dependencies]
click = "^8.1.3"

[[tool.poetry.source]]
name = "piwheels"
url = "https://www.piwheels.org/simple"
```

Given a secondary source, `click` and friends will be locked from the default source (PyPI):
```toml
[tool.poetry.dependencies]
click = "^8.1.3"

[[tool.poetry.source]]
name = "piwheels"
url = "https://www.piwheels.org/simple"
secondary = true
```

Skimming the current code, I'm actually a bit confused on how this is implemented, but it is what we currently do.

Reply:
Indeed, salting some prints into `add_repository` yields the following: 
```
piwheels 0
pypi 1
[<poetry.repositories.legacy_repository.LegacyRepository object at 0x7fdebc55dae0>, <poetry.repositories.pypi_repository.PyPiRepository object at 0x7fdebc5be3e0>]
```

I do think that the logic is tied into a hard to decipher pretzel, and the `RepositoryPriorities` enum in this PR needs to change.

Reply:
In a discussion on Discord we came to the conclusion that

> * First, "primary" sources are considered in the order they are defined. These are entries in the pyproject.toml without `default = true` or `secondary = true`. This is equivalent to pip's `--extra-index-url`.
> 
> * Second, the `default` source is considered. If the default source is not overidden in pyproject.toml, this is PyPI. This is equivalent to pip's `--index-url`.
> 
> * Finally, all `secondary` sources are considered in the order they are defined. This is a Poetry invention.

describes how it should be. However, it isn't the current behavior. So let's drop the documentation for now, do only the refactoring here and add documentation + bugfix to comply with the description above in a follow-up PR because it's easier to do after the refactoring than before.

Sorry for the back and forth. We really appreciate your contribution. 🙏

Reply:
Ha don't worry about it! It took me quite some time to get familiar with enough code to work on this and the PyPI insertion and default/primary/secondary sources got me confused -- glad to see I'm not the only one. I'd be more than happy to change the behavior into something that can be explained easily (but I figured that a pure refactor should not change the current behavior). I'll take this up in my attempt at https://github.com/python-poetry/poetry/pull/5984.